Bluetooth: Host: Clean up buffer user data in case of error

hermabe opened the pull request:
`bt_l2cap_chan_send_cb()` overwrote the buffer user data for internal
use. In the case where sending fails, this would be visible for the
caller. If the caller relied on the buffer user data to be unchanged,
this could cause unexpected behavior.

L2CAP tx metadata was also not freed in the error case.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe committed:
Bluetooth: Host: Clean up buffer user data in case of error

`bt_l2cap_chan_send_cb()` overwrote the buffer user data for internal
use. In the case where sending fails, this would be visible for the
caller. If the caller relied on the buffer user data to be unchanged,
this could cause unexpected behavior.

L2CAP tx metadata was also not freed in the error case.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

jori-nordic commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -2952,6 +2952,7 @@ int bt_l2cap_chan_send_cb(struct bt_l2cap_chan *chan, struct net_buf *buf, bt_co
 {
 	struct bt_l2cap_le_chan *le_chan = BT_L2CAP_LE_CHAN(chan);
 	struct l2cap_tx_meta_data *data;
+	void *old_user_data = l2cap_tx_meta_data(buf);
```
can you add a check in one of the bsim tests for this? so that it fails without this patch.

hermabe committed:
Bluetooth: Test: Add test to verify L2CAP sending not changing user data

Regression test for an issue where in cases where sending a buffer over
L2CAP failed, the buffer user data was not restored.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

stephanosio commented:
Rebased to fix CI failure
